src/async/csharp/Program.cs

```diff
@@ -0,0 +1,141 @@
+    connection.EnableAlerts();
+
+    // There is no async GetDevice because it just instantiates a device object
+    // without communicating with anything.
+    var xDevice = connection.GetDevice(X_DEVICE_ADDRESS);
+
+    // Everywhere you await a Zaber async function call, you have the option to do some other
+    // processing in parallel. Instead of using "await" immediately, you can assign the function's
+    // return value to a Task variable and then await it later after doing something else. You can
+    // also use Task.WhenAll() to wait for multiple async operations to complete.
+    var identifyTask = xDevice.IdentifyAsync();
+    // Could do something else here.
+    await identifyTask; // Block until IdentifyAsync() completes.
+    var xAxis = xDevice.GetAxis(X_AXIS_NUMBER);
+
+    Axis yAxis;
+    if (Y_DEVICE_ADDRESS == X_DEVICE_ADDRESS)
+    {
+        yAxis = xDevice.GetAxis(Y_AXIS_NUMBER);
+    }
+    else
+    {
+        var yDevice = connection.GetDevice(Y_DEVICE_ADDRESS);
+        await yDevice.IdentifyAsync();
+        yAxis = yDevice.GetAxis(Y_AXIS_NUMBER);
+    }
+
+    // Home the devices and wait until done.
+    // This is an example of overlapping async commands. Order of execution is not
+    // guaranteed and with more than two or three there is a risk of a device error.
+    await Task.WhenAll(xAxis.HomeAsync(), yAxis.HomeAsync());
+
+    // Grid scan loop
+    Axis[] axes = [xAxis, yAxis];
+    await foreach (var coords in Grid(X_GRID_POINTS, Y_GRID_POINTS))
+    {
+        // Here's one way of controlling the devices in parallel. This loop potentially sends the
+        // move commands and waits for the acknowledgements on different threads. There is a
+        // risk of overlapping too many commands this way; it's only safe here because this
+        // example only controls two axes. This pattern is not appropriate for more than three;
+        // use a non-parallel loop in such cases.
+        await Parallel.ForEachAsync(Enumerable.Range(0, axes.Length), async (index, _) =>
+        {
+            var position = coords[index] * GRID_SPACING;
+            // Avoid the tempation to save the tasks in an array and await them as a group in the
+            // case of movement commands, as the device may not be able to consume the commands as
+            // fast as you can send them, and a system error may occur. Two or three should
+            // be safe, as in the home command above.
+            // Again, you can move the await to a later point in the loop if you want to.
+            await axes[index].MoveAbsoluteAsync(position, GRID_SPACING_UNITS, false);
+        });
+
+        // At this point the axes are moving and we may have many milliseconds or seconds to
+        // do other work (not shown).
+
+        // Now when we need to be sure the devices have stopped moving, we can wait until they are idle.
+        foreach(var axis in axes)
+        {
+            await axis.WaitUntilIdleAsync();
```

Comment:
This is not a good example. This may as well be synchronous. Since it's doing the wait_until_idle.